diff -Naur connman-0.77/Makefile.am connman-0.77.patch/Makefile.am
--- connman-0.77/Makefile.am	2011-08-23 03:16:04.000000000 +0200
+++ connman-0.77.patch/Makefile.am	2011-11-17 04:41:54.624794205 +0100
@@ -98,7 +98,7 @@
 
 scriptdir = $(libdir)/connman/scripts
 
-storagedir = $(localstatedir)/lib/connman
+storagedir = /run/cache/connman
 
 configdir = ${sysconfdir}/connman
 
diff -Naur connman-0.77/Makefile.in connman-0.77.patch/Makefile.in
